### PR TITLE
refactor type test generation

### PR DESCRIPTION
**Pull request by o-**

### CODE REVIEW

File: rir/src/compiler/opt/type_speculation.cpp [hunk 1]
```diff
@@ -16,106 +17,64 @@ void TypeSpeculation::apply(RirCompiler&, ClosureVersion* function,
...
-                    }
-                }
-            }
-            ip = next;
+        }
+
+        if (!speculateOn)
+            return;
+
+        if (auto cp = checkpoint.next(i)) {
```
Comment (JanJecmen) [inline rir/src/compiler/opt/type_speculation.cpp]: For the case, this used to be `auto cp = checkpoint.at(force)` and not `next`

Comment (o-) [reply rir/src/compiler/opt/type_speculation.cpp]: I see. guess `checkpoint.next(speculateOn)` makes the most sense, doesn't it?

Comment (o-) [reply rir/src/compiler/opt/type_speculation.cpp]: the reason I am working on this is, that I saw in matrix-for we added a sub-par test and I wanted to add more types.

so I started with this refactoring, that maybe should have been done earlier...

Comment (JanJecmen) [reply rir/src/compiler/opt/type_speculation.cpp]: Good point.. My reasoning was to not have the checkpoint and the assume too far apart, but I guess this allows to hoist the assume further.

By more types you mean something like `(logical | real)'`?

Comment (charig) [reply rir/src/compiler/opt/type_speculation.cpp]: I wonder whether `checkpoint.next(speculateOn)` could have as a consequence that less checkpoints are available

**Pull request merged**